src/main/java/org/opensearch/index/store/directio/CryptoDirectIODirectory.java

```
@@ -196,7 +218,35 @@ public void deleteFile(String name) throws IOException {
                 LOGGER.warn("Failed to get file size", e);
             }
         }
-
+        EncryptionCache.getInstance().invalidateFile(filePath);
         super.deleteFile(name);
     }
+
+    /**
+     * Calculate content length with OSEF validation.
+     * Fast path: check cache first to avoid FileChannel open.
+     */
+    private long calculateContentLengthWithValidation(Path file, long rawFileSize) throws IOException {
+        if (rawFileSize < EncryptionMetadataTrailer.MIN_FOOTER_SIZE) {
+            return rawFileSize;
+        }
+
+        // Fast path: check cache first - avoids FileChannel open
+        String filePath = file.toAbsolutePath().toString();
+        EncryptionFooter cachedFooter = EncryptionCache.getInstance().getFooter(filePath);
+        if (cachedFooter != null) {
+            return rawFileSize - cachedFooter.getFooterLength();
+        }
+
+        // Slow path: read footer from disk
+        try (FileChannel channel = FileChannel.open(file, StandardOpenOption.READ)) {
+            try {
+                EncryptionFooter footer = EncryptionFooter.readFromChannel(file, channel, dataKeyBytes);
+                return rawFileSize - footer.getFooterLength();
+            } catch (EncryptionFooter.NotOSEFFileException e) {
+                LOGGER.debug("Not an OSEF file: {}", file);
+                return rawFileSize;
```

let's not swallow the exception.

---

removed